fix: show code excerpt errors inline

attfarhan opened the pull request:
Fixes https://github.com/sourcegraph/sourcegraph/issues/1987. 

![image](https://user-images.githubusercontent.com/16265452/53126427-f191e800-3514-11e9-813f-3873e9ee7991.png)

felixfbecker commented on shared/src/components/CodeExcerpt.tsx:
```diff
@@ -106,40 +102,43 @@ export class CodeExcerpt extends React.PureComponent<Props, State> {
+                    {this.state.blobLinesOrError && isErrorLike(this.state.blobLinesOrError) && (
+                        <div
+                            className="alert alert-danger "
+                            dangerouslySetInnerHTML={{ __html: this.state.blobLinesOrError.message }}
```
Error messages are not HTML

attfarhan commented on shared/src/components/CodeExcerpt.tsx:
```diff
@@ -2,8 +2,9 @@ import { range } from 'lodash'
 import VisibilitySensor from 'react-visibility-sensor'
 import { combineLatest, Observable, Subject, Subscription } from 'rxjs'
-import { filter, switchMap } from 'rxjs/operators'
+import { catchError, filter, switchMap } from 'rxjs/operators'
```
test

attfarhan commented on shared/src/components/CodeExcerpt.tsx:
```diff
@@ -2,8 +2,9 @@ import { range } from 'lodash'
 import { combineLatest, Observable, Subject, Subscription } from 'rxjs'
-import { filter, switchMap } from 'rxjs/operators'
+import { catchError, filter, switchMap } from 'rxjs/operators'
 import { highlightNode } from '../util/dom'
```
test

slimsag commented:
It would be really nice to get a fix for this merged in for 3.2!

attfarhan commented:
I'll revisit this, thanks for pinging.

attfarhan commented:
I updated this to look like so:

![image](https://user-images.githubusercontent.com/16265452/54397384-85b62180-4673-11e9-8984-2713d0b3d05c.png)

![image](https://user-images.githubusercontent.com/16265452/54397470-cdd54400-4673-11e9-91f9-a3d6f21d23d6.png)


I didn't implement exactly the design requested, but I think this updated screenshot looks totally fine, and it's an improvement for us to show the error message at all anyways. 

Anyways, I think this is actually gonna be a very rare case now, since we fixed the issue which caused some results to be blank: https://github.com/sourcegraph/sourcegraph/issues/2615.

attfarhan commented:
@francisschmaltz new UI:

![image](https://user-images.githubusercontent.com/16265452/54398637-1a228300-4678-11e9-9382-960b9bba9c8f.png)

![image](https://user-images.githubusercontent.com/16265452/54398616-0c6cfd80-4678-11e9-8817-0ec44bec6fe4.png)

felixfbecker commented:
Two comments:
- I find it inconsistent that in every error display we use a red background, but not in this one.
- As mentioned in my previous review, it's misleading that the font is still monospace. Since in the non-error case, code appears in this place (which is also monospace and even colored) it's not obvious that these are error messages. We don't use monospace for errors elsewhere, so the association we built is monospace font = code content.

francisschmaltz commented:
> monospace font

Totally agree. Font should be our standard stack. 

> background

With the card UI it looks really bad. This is a stop gap to get us to an acceptable state. More on this should be assigned in 3.3 forward.

francisschmaltz commented:
> Having any display of errors in 3.2 is strictly more important than blocking this PR on any style, font, etc.  even if that means merging as-is and delaying those changes until 3.3.

As long as design is a hard-requirement for 3.3 we can do this - but the changes I request above should land first. We should have a better discussion in 3.3 and define ownership.

slimsag commented:
@francisschmaltz @felixfbecker Here is how I measure this:

**Today:**

1. You are not told about errors at all today (very bad).
2. The UX displays an empty box where errors should be displayed (very ugly).
3. Customers are running into this and we have to ask them to open the chrome console to find the error (very bad).

**If we merge this PR:**

1. You are told about errors (good).
2. The UX does not display empty boxes when errors occur (good).
3. We don't have to ask people to open the chrome console to tell us error messages (good).
4. The UI style doesn't live up to our most ideal standard (bad).

We're changing 3 negatives into 3 positives & 1 negative.

We should review changes based on their long term benefit for the good of our users, not based on whether or not they are perfect (most changes are not). Development is incremental, and doing everything all at once is simply not possible. This is a fundamental part of writing software. At some point, we have to say that `havingThisChange > !havingThisChange`.

For the sake of our users, and for the reasons above, I am stating that point is _now_.

I am merging this PR, if anyone disagrees with this decision I encourage you to send a PR to revert the change and make a strong user-driven argument for why we should do so.

@attfarhan should follow up on all style comments here by filing issues and prioritizing them as appropriate.
